### PR TITLE
implement 2D map and basic operations

### PR DESCRIPTION
**Pull request by imrn99**
# implement 2D map and basic operations

Implementation of 2D combinatorial maps.

To-Do before merging:

- [x] 2-map editing interfaces
- [x] first implementation for orbits / i-cells computation
- [x] Complete in-code documentation
- [x] (optional) Add basic testing: as doc examples

## Scope

- [x] Code: addition to `honeycomb-core`

## Type of change

- [x] New feature(s)

## Other

...

## Necessary follow-up

- [x] Needs documentation: in user guide, mostly
- [x] Needs testing
- [x] Other: Implement proper I/O and conversion code to mesh structures

## Mentions

...

### CODE REVIEW

File: honeycomb-core/src/cmap/embed.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+/// println!("dart {} associated cells: {:#?}", darts[1].id(), embedded_cells[darts[1].id()]);
+/// ```
+///
+pub struct DartCells {
+    /// Vertex unique identifier.
+    pub vertex_id: usize,
+    /// Face unique identifier.
+    pub face_id: usize,
+    /// Volume unique identifier.
+    pub volume_id: usize,
```
Comment (cedricchevalier19) [inline honeycomb-core/src/cmap/embed.rs:46]: No custom type for these ids ?

File: honeycomb-core/src/cmap/map.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    cells: Vec<DartCells>,
+    /// List of darts composing the map.
+    ///
+    /// Used to ...
+    darts: Vec<Dart>,
+    /// Array representation of the beta functions.
+    ///
+    /// This should eventually be replaced by a better
+    /// structure, supported by benchmarking.
+    betas: Vec<[usize; 2]>,
```
Comment (cedricchevalier19) [inline honeycomb-core/src/cmap/map.rs:43]: Should be `DartIdentifier` ?

**Review by cedricchevalier19 — COMMENTED** [review]
I think identifiers should be custom types.  It can help to know what kind of indirections are stored.

We can change implementation from simple aliasing to more opaque type.

**Comment by imrn99** [comment]
> I think identifiers should be custom types. It can help to know what kind of indirections are stored.
> 
> We can change implementation from simple aliasing to more opaque type.

I replaced `usize` with custom ID types & updated methods accordingly in 971cf6ab75d7cfa0d580299acc1ed0495cb43257

Side effect: when the custom types are set as `usize`, many warnings pop up because of now useless type conversions. Overall, this still is better this way since it will help differentiate `Dart` and `DartIdentifier` in the long run.

File: honeycomb-core/src/twomap.rs [hunk 3]
```diff
@@ -0,0 +1,1082 @@
...
+        rhs_dart_id: DartIdentifier,
+        policy: SewPolicy,
+    ) {
+        macro_rules! stretch {
+            ($replaced: expr, $replacer: expr) => {
+                self.dart_data.associated_cells[$replaced as usize].vertex_id =
+                    self.dart_data.associated_cells[$replacer as usize].vertex_id
+            };
+        }
+        // --- topological update
```
Comment (cedricchevalier19) [inline honeycomb-core/src/twomap.rs:621]: This macro is also used lines 698-703. Do you want to keep the duplication?

File: honeycomb-core/src/twomap.rs [hunk 4]
```diff
@@ -0,0 +1,1082 @@
...
+    /// - `lhs_dart_id: DartIdentifier` -- ID of the dart to separate.
+    /// - `policy: UnsewPolicy` -- Geometrical unsewing policy to follow.
+    ///
+    /// Note that we do not need to take two darts as arguments since the
+    /// second dart can be obtained through the *β<sub>1</sub>* function. The
+    /// *β<sub>0</sub>* function is also updated.
+    ///
+    /// # Return / Panic
+    ///
+    /// The method may panic if *I* is neither 1 or 2.
```
Comment (cedricchevalier19) [inline honeycomb-core/src/twomap.rs:856]:
```suggestion
    /// The method may panic if *I* is neither 1 nor 2.
```

File: honeycomb-core/src/twomap.rs [hunk 5]
```diff
@@ -0,0 +1,1082 @@
...
+    ///
+    /// - `lhs_dart_id: DartIdentifier` -- ID of the dart to separate.
+    /// - `policy: UnsewPolicy` -- Geometrical unsewing policy to follow.
+    ///
+    /// Note that we do not need to take two darts as arguments since the
+    /// second dart can be obtained through the *β<sub>2</sub>* function.
+    ///
+    /// # Return / Panic
+    ///
+    /// The method may panic if *I* is neither 1 or 2.
```
Comment (cedricchevalier19) [inline honeycomb-core/src/twomap.rs:901]:
```suggestion
    /// The method may panic if *I* is neither 1 nor 2.
```

**Review by cedricchevalier19 — APPROVED** [review]
Good work but try to keep PR smaller next time!

**Pull request merged**